### PR TITLE
Provide torch Dataset using its IterableDataset interface

### PR DESCRIPTION
**Pull request by eddyxu**
Pytorch has merged a shuffler that works with IterableDataset, we can limit the scope of `rikai.torch.data` to a Dataset instead. It can leave multi-process loading and shuffling to `pytorch` to handle

### CODE REVIEW

**Comment by eddyxu** [comment]
When it uses with `torch.utils.data.DataLoader(num_workers>0)` , and the partitions are not _perfectly_ balanced,  data loader hangs.

**Comment by changhiskhan** [comment]
do you want me to review yet? or wait til you finish the work?

**Comment by eddyxu** [comment]
Please wait a bit. I have not figured out what was wrong yet.

**Comment by eddyxu** [comment]
Maybe relevant https://github.com/pytorch/pytorch/issues/15808

**Comment by eddyxu** [comment]
It seems that a few partitions in the dataset returning `_IterableDatasetStopIteration` and then the full dataset stops. 

```
```

File: python/tests/conftest.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
 
 # Third Party
 import pytest
+from torch.utils.data import DataLoader  # Prevent DataLoader hangs
```
Comment (changhiskhan) [inline python/tests/conftest.py:18]: why does it hang if you don't import it here?

File: python/tests/parquet/test_shuffler.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
 #  limitations under the License.
 
 
+from random import shuffle
```
Comment (changhiskhan) [inline python/tests/parquet/test_shuffler.py:16]: unused?

File: python/tests/torch/test_data.py [hunk 3]
```diff
@@ -18,13 +18,18 @@
...
 from pyspark.sql import Row, SparkSession
+from torch.utils.data import DataLoader as torchDataLoader
 
 # Rikai
 from rikai.numpy import wrap
-from rikai.torch import DataLoader
+from rikai.torch import DataLoader, Dataset
 from rikai.types import Box2d, Image
 
+torch.multiprocessing.set_sharing_strategy("file_system")
```
Comment (changhiskhan) [inline python/tests/torch/test_data.py:31]: consider putting this in a session scoped fixture or inside pytest_sessionstart ?

File: python/rikai/torch/data.py [hunk 4]
```diff
@@ -13,19 +13,90 @@
...
 
 # Third Party
 import numpy as np
+import torch
+from torch.utils.data import IterableDataset
 
 # Rikai
 from rikai.mixin import ToNumpy
-from rikai.parquet.dataset import Dataset
+from rikai.parquet.dataset import Dataset as pgDataset
```
Comment (changhiskhan) [inline python/rikai/torch/data.py:30]: why `pg`?

**Review by changhiskhan — APPROVED** [review]

File: python/rikai/torch/data.py (hunk 4 above)
Comment (eddyxu) [reply python/rikai/torch/data.py:30]: hmm should be `pq`... for `parquet`

File: python/tests/conftest.py (hunk 1 above)
Comment (eddyxu) [reply python/tests/conftest.py:18]: I have no good theory, need to talk with the team.
But TLDR after days of trying is that `torch.DataLoader` needs to be imported before` pyspark.sql`

File: python/tests/parquet/test_shuffler.py (hunk 2 above)
Comment (eddyxu) [reply python/tests/parquet/test_shuffler.py:16]: done

File: python/tests/torch/test_data.py (hunk 3 above)
Comment (eddyxu) [reply python/tests/torch/test_data.py:30]: I am not very familar with the convention for pytest fixture.

Should this be:
```
@pytest.fixture(scope="session"):
def torch_session():
    torch.multiprocessing.set_sharing_strategy("file_system")
```

and let each of the torch test to declare this parameter?

File: python/rikai/parquet/dataset.py [hunk 5]
```diff
@@ -18,6 +18,7 @@
 # Standard Library
 import importlib
 from typing import Any, Dict, List, Optional
+from _pytest.fixtures import yield_fixture
```
[the remark this answers is not in the document]
Comment (eddyxu) [reply python/rikai/parquet/dataset.py]: Removed.

File: python/tests/conftest.py (hunk 1 above)
Comment (eddyxu) [reply python/tests/conftest.py:18]: Added a `rikai/tests/torch/conftest.py`

File: python/tests/torch/test_data.py (hunk 3 above)
Comment (eddyxu) [reply python/tests/torch/test_data.py:30]: Added a fixture in `python/tests/torch/conftest.py`
